mat64/io.go

```diff
@@ -114,6 +150,69 @@ func (m *Dense) UnmarshalBinary(data []byte) error {
 	return nil
 }
 
+// UnmarshalBinaryFrom decodes the binary form into the receiver and returns
+// the number of bytes read and an error if any.
+// It panics if the receiver is a non-zero Dense matrix.
+//
+// See MarshalBinary for the on-disk layout.
+//
+// Limited checks on the validity of the binary input are performed:
+//  - matrix.ErrShape is returned if the number of rows or columns is negative,
+//  - an error is returned if the resulting Dense matrix is too
+//  big for the current architecture (e.g. a 16GB matrix written by a
+//  64b application and read back from a 32b application.)
+// UnmarshalBinary does not limit the size of the unmarshaled matrix, and so
+// it should not be used on untrusted data.
+func (m *Dense) UnmarshalBinaryFrom(r io.Reader) (int, error) {
+	if !m.isZero() {
+		panic("mat64: unmarshal into non-zero matrix")
+	}
+
+	var (
+		n   int
+		buf [8]byte
+	)
+	nn, err := readFull(r, buf[:])
+	n += nn
+	if err != nil {
+		return n, err
+	}
+	rows := int64(binary.LittleEndian.Uint64(buf[:]))
+
+	nn, err = readFull(r, buf[:])
+	n += nn
+	if err != nil {
+		return n, err
+	}
+	cols := int64(binary.LittleEndian.Uint64(buf[:]))
+	if rows < 0 || cols < 0 {
+		return n, errBadSize
+	}
+
+	size := rows * cols
+	if int(size) < 0 || size > maxLen {
+		return n, errTooBig
+	}
+
+	m.mat.Rows = int(rows)
```

Comment:
Can this whole stanza be replaced with m.reuseAs(int(rows), int(cols))?

The same thing in UnmarshalBinary.